jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPathsIterator.java

```diff
@@ -0,0 +1,283 @@
+     * @param source the source/from/start vertex
+     * @param sink the sink/to/end vertex
+     * @param spFactory the factory function to create new instances of the shortest path algorithm.
+     */
+    public YenKShortestPathsIterator(
+        Graph<V, E> graph, V source, V sink,
+        Function<Graph<V, E>, ShortestPathAlgorithm<V, E>> spFactory)
+    {
+        this.spFactory = spFactory;
+        this.graph = graph;
+        this.sink = sink;
+        a = new ArrayList<GraphPath<V, E>>();
+        // Determine the shortest path from the source to the sink.
+        ShortestPathAlgorithm<V, E> shortestPathAlgorithm = spFactory.apply(graph);
+        GraphPath<V, E> p = shortestPathAlgorithm.getPath(source, sink);
+        if (null != p) {
+            a.add(p);
+        }
+        nextIndex = 0;
+        // // Initialize the heap to store the potential kth shortest paths.
+        b = new PriorityQueue<GraphPath<V, E>>(new Comparator<GraphPath<V, E>>()
+        {
+            @Override
+            public int compare(GraphPath<V, E> arg0, GraphPath<V, E> arg1)
+            {
+                return Double.compare(arg0.getWeight(), arg1.getWeight());
+            }
+        });
+    }
+
+    /**
+     * Construct the object for searching k shortest paths in provided graph using
+     * {@link DijkstraShortestPath} as the base algorithm.
+     * 
+     * @param graph the underlying graph on which the paths are searched
+     * @param source the source/from/start vertex
+     * @param sink the sink/to/end vertex
+     */
+    public YenKShortestPathsIterator(Graph<V, E> graph, V source, V sink)
+    {
+        this(graph, source, sink, DijkstraShortestPath<V, E>::new);
+    }
+
+    @Override
+    public boolean hasNext()
+    {
+        if (nextIndex < a.size()) {
+            return true;
+        }
+        tryAddNext();
+        return nextIndex < a.size();
+    }
+
+    @Override
+    public GraphPath<V, E> next()
+    {
+        if (nextIndex < a.size()) {
+            return a.get(nextIndex++);
+        }
+        tryAddNext();
+        if (nextIndex >= a.size()) {
+            throw new NoSuchElementException();
+        }
+        return a.get(nextIndex++);
+    }
+
+    /**
+     * Determine and store in {@link #b} the next shortest paths and determine the actual shortest
+     * path and add it to {@link #a}. Instead of altering the underlying graph, this implementation
+     * masks edges and nodes that shall not be considered using {@link MaskSubgraph}.
+     */
+    private void tryAddNext()
+    {
+        GraphPath<V, E> lastShortestPath = a.get(a.size() - 1);
+        // The spur node ranges from the first node to the next to last
+        // node in the previous k-shortest path.
+        for (int i = 0; i < lastShortestPath.getLength() - 1; i++) {
+            // Spur node is retrieved from the previous k-shortest path
+            V spurNode = lastShortestPath.getVertexList().get(i);

```

Comment:
This unfortunately is a bit tricky. A Path in a graph can be unambiguously expressed in terms of a sequence of edges. A path in a graph expressed by a sequence of vertices may however be ambiguous. For instance, in a multigraph where there may be multiple edges between the same pair of vertices, a path must be expressed in terms of the edges. Therefore, most shortest path algorithms will return a GraphPath expressed in terms of the edges of the path. Each invocation to getVertexList() may recompute the path expressed in terms of vertices. This gets potentially very expensive, since you invoke getVertexList() very often. It would be better to cache the result of this method. To understand what I mean, have a look at the implementation of GraphPath and GraphWalk.

Reply:
This is not ambiguous for Yen's algorithm, as Yen's algorithm will not work if there are multiple edges between the same vertices.

Reply:
Indeed you need to use some caching here as method `getVertexList()` might reconstruct a new list on each invocation. Take a look at the internals of `GraphWalk` to understand this.

Reply:
I am storing now both the GraphPath as well as the call to getVertexList. IMO this is BAD DESIGN - it fixes a problem at the wrong place. 
If an implementer of GraphPath considers its implementation of getVertexList to be causing performance issues, that implementer of GraphPath should keep a weak reference to a once computed vertex list and return it in subsequent calls. 
Currently only GraphWalk implements GraphPath and it has already slots for both vertexList as well as edgeList, any caching logic should be added here.
Users of GraphPath should not be concerned with caching results.

Reply:
I've been working on some optimizations for graphwalk, including this caching issue. Will be submitted soon.